components/raftstore/src/store/fsm/apply.rs

```diff
@@ -1094,7 +1106,13 @@ where
 
             match res {
                 ApplyResult::None => {}
-                ApplyResult::Res(res) => results.push_back(res),
+                ApplyResult::Res(res) => {
+                    results.push_back(res);
+                    if self.wait_data {
+                        apply_ctx.committed_count -= committed_entries_drainer.len();
+                        break;
+                    }
```

Comment:
please add some comments on this block.
Question: why if it's wait_data, it breaks the loop. 
One of my confusion is when witness is converted to non-witness, does it have any difference with the scenario a new peer is created and needs to catch up the data by snapshot.

Reply:
> please add some comments on this block.

Sure, will add after all your questions have been confirmed.

>  Question: why if it's wait_data, it breaks the loop.

If it's witness before, but a command changes it to non-witness, and starts applying all following command. It can cause problem if the command depends on previous side effect. 

> One of my confusion is when witness is converted to non-witness, does it have any difference with the scenario a new peer is created and needs to catch up the data by snapshot.

The difference is that the new peer is `uninitialized` before `apply snapshot`, and applied `non-witness switch` makes `apply thread` thinks it is already `non-witness`, but it is not. So it should not continue to apply subsequent logs until the `apply snapshot` is complete.